### DISCUSSION TITLE
Am I overlooking documentation for midi?

### DISCUSSION THREAD

**Discussion by RyanParsley** [cardonabits/haxo-rs#21] [Q&A] [answered]
I like plugging my haxophone in and making noise with it, but midi out feels like a fun thing to play with too. I see a [closed issue around supporting midi](https://github.com/cardonabits/haxo-rs/issues/14), but I don't see anything in the main readme. 

When I poke around the source code, I see a script folder with a handful of scripts that don't seem to be documented. 

I attempted to run the create-midi-gadget.sh and everything seems to do what it's supposed to except `modprobe libcomposite` doesn't have any output and  `echo $(ls /sys/class/udc) > UDC` seems to be piping nothing into the UDC file. Is there maybe a dependency to install before this script can succeed?

Related to docs, but not directly midi related, It seems like the leveraging haxo.service is the way to go, but project readme suggests an alternate route. What's the happy path here? Am I simply overlooking some documentation?

**Comment by jcard0na** [comment]
Hi @RyanParsley,

If you are using an image from https://bitly.com/rpi-zero that you download after 12/29/2023, USB MIDI should work out of the box.
If you are building your own images, it seems like you are missing step 5 [from here](https://github.com/cardonabits/haxo-hw/blob/main/docs/development.md#initial-configuration):

> 5. [RP Zero Only] Enable device-mode USB driver (edit /boot/config.txt to add)
> 
```
 [all]
 dtoverlay=dwc2
```
> 
> and add dwc2 to /etc/modules

I'm almost certain that this is the problem you have because I encountered exactly the same problem with an empty `/sys/class/udc`.  I had added `dtoverlay=dwc2` to a different section of `config.txt` (`[cm4]`) which resulted in the `dtoverlay` statement being ignored.

Let me know if that helps!

**Reply by RyanParsley** [reply]
I see the missing connection. I was in the haxo-rs readme and needed some context from the hw docs. I had worked through these docs in the past, but before you added the dwc2 bit. This time around, as I was updating, I overlooked this difference. Thank you! I now see the haxophone as a midi device in Ableton Live when it boots up. I'm failing to see activity in live when I play the haxophone, but this feel like progress and I'll dig into the docs to see if I simply missed another step. 

Thank so much for your quick reply and for sharing this fun project!

**Reply by petermoz** [reply, marked as answer]
Hi @RyanParsley,

If you're building from source code, the other thing that might trip you up is that the `midi` feature is not enabled by default. You have to do `cargo build --release --features midi`. Without that, the haxophone will show up as a midi device, but not generate any midi output.

@jcard0na do you think `midi` should be made a default feature?

**Reply by jcard0na** [reply]
I made it a feature because it requires USB On The Go, which I believe is only available on the Pi Zeros.  The Pi 3 and 4 only have USB host, and I don't know exactly how badly they would fail if haxo001 attempted a MIDI connection.

**Reply by petermoz** [reply]
Ah ok makes sense. Glad I bought a Pi Zero then :)

**Reply by RyanParsley** [reply]
I had missed the change to the boot.txt steps. Now that I've built with the midi feature flag everything is working as I'd expect. Thank you both for helping me fill in the gaps.
